File: medium/2975/main.go
package main

import "sort"

func main() {

}

// Solution from AkitKumar on leetcode
// https://leetcode.com/problems/maximum-square-area-by-removing-fences-from-a-field/solutions/7498260/find-biggest-square-from-fence-gaps-begi-8o9m/
func AkitKumar(m int, n int, hFences []int, vFences []int) int {
	const MOD = 1000000007

	prep := func(cuts []int, limit int) []int {
		sort.Ints(cuts)
		out := []int{1}
		out = append(out, cuts...)
		out = append(out, limit)
		return out
	}

	h := prep(hFences, m)
	v := prep(vFences, n)

	gapSet := make(map[int]bool)

	for i := 0; i < len(h); i++ {
		for j := i + 1; j < len(h); j++ {
			gapSet[h[j]-h[i]] = true
		}
	}

	best := 0
	for i := 0; i < len(v); i++ {
		for j := i + 1; j < len(v); j++ {
			d := v[j] - v[i]
			if d > best && gapSet[d] {
				best = d
			}
		}
	}

	if best == 0 {
		return -1
	}
	return int((int64(best) * int64(best)) % MOD)
}

// Solution from PattoBear on leetcode
// https://leetcode.com/problems/maximum-square-area-by-removing-fences-from-a-field/solutions/7498442/golang-good-ass-hints-by-pattobears-mq85/
func PattoBear(m int, n int, hFences []int, vFences []int) int {
	hFences = append(hFences, 1, m)
	vFences = append(vFences, 1, n)
	hLen := len(hFences)
	vLen := len(vFences)
	mod := int(1e9 + 7)
	diffSet := map[int]struct{}{}

	for i := range hLen - 1 {
		for j := i + 1; j < hLen; j++ {
			diffSet[abs(hFences[i]-hFences[j])] = struct{}{}
		}
	}

	out := -1

	for i := range vLen - 1 {
		for j := i + 1; j < vLen; j++ {
			side := abs(vFences[i] - vFences[j])
			if _, ok := diffSet[side]; ok {
				out = max(out, side*side)
			}
		}
	}

	return out % mod
}

func abs(n int) int {
	if n > 0 {
		return n
	}
	return -n
}

File: medium/2975/main.py
# 2975. Maximum Square Area by Removing Fences From a Field
# https://leetcode.com/problems/maximum-square-area-by-removing-fences-from-a-field/description

MOD = 10**9 + 7


def akit_kumar(m: int, n: int, h_fences: list[int], v_fences: list[int]) -> int:
    """
    Solution from AkitKumar on leetcode
    https://leetcode.com/problems/maximum-square-area-by-removing-fences-from-a-field/solutions/7498260/find-biggest-square-from-fence-gaps-begi-8o9m/
    """

    def prep(cuts: list[int], limit: int) -> list[int]:
        return sorted([1] + cuts + [limit])

    h = prep(h_fences, m)
    v = prep(v_fences, n)

    gap_set = set()
    for i in range(len(h)):
        for j in range(i + 1, len(h)):
            gap_set.add(h[j] - h[i])

    best = 0
    for i in range(len(v)):
        for j in range(i + 1, len(v)):
            d = v[j] - v[i]
            if d > best and d in gap_set:
                best = d

    if best == 0:
        return -1
    return (best * best) % MOD


def patto_bear(m: int, n: int, h_fences: list[int], v_fences: list[int]) -> int:
    """
    Solution from PattoBear on leetcode
    https://leetcode.com/problems/maximum-square-area-by-removing-fences-from-a-field/solutions/7498442/golang-good-ass-hints-by-pattobears-mq85/
    """
    h_fences = h_fences + [1, m]
    v_fences = v_fences + [1, n]

    diff_set = set()
    for i in range(len(h_fences) - 1):
        for j in range(i + 1, len(h_fences)):
            diff_set.add(abs(h_fences[i] - h_fences[j]))

    out = -1
    for i in range(len(v_fences) - 1):
        for j in range(i + 1, len(v_fences)):
            side = abs(v_fences[i] - v_fences[j])
            if side in diff_set:
                out = max(out, side * side)

    return out % MOD if out != -1 else -1
